### PR TITLE
1.x: Datepicker changes

### PR DESCRIPTION
**Pull request by c-frater**
## Description

This change is in preparation for the 2.0.0 release to allow for an easier migration

Updates the date picker to allow an `input` as the element to attach the datepicker to as well as the following features under a `v2Features` flag:
- Only modify the typed date value when it is valid

The following options have been added:
- `allowViewChange`: toggles the ability to switch to year/decade view
- `dateValidityCallback`: Allows external validation of the `input` element

The tests have also been converted to use a headless chrome instance instead of PhantomJS as there were multiple issues caused by the feature support within PhantomJS on multiple operating systems.

## Motivation and Context

To increase support when using Hexagon alongside other UI frameworks (such as Bootstrap)

## How Has This Been Tested?

Added new tests

## Screenshots (if appropriate):
screen shot 2018-11-09 at 15 15 11

## Checklist:

- [x] I have added a tag to this pull request that indicates the impact of the change (patch, minor or major)
- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly. (This includes updating the changelog).
- [x] I have read the **CONTRIBUTING** document.
- [x] All my changes are covered by tests.
- [x] This request is ready to review and merge

### CODE REVIEW

File: build/main/karma.js [hunk 1]
```diff
@@ -25,11 +25,12 @@ var flatten = require('flatten')
 // runs a set of testPackages in karma (a test package is an object
 // that looks like: { name, css, dependencesJs, moduleJs, specJs }
 // destDir is a directory where the results should be written to
-module.exports = function (testPackages, destDir, phantomOnly) {
+module.exports = function (testPackages, destDir, chromeHeadlessOnly) {
   return fs.copyAsync(path.join(__dirname, 'test-helpers.js'), path.join(destDir, 'lib', 'test-helpers.js')).then(function() {
     return Promise.all(testPackages.map(function (testPackage) {
       var moduleDir = path.join(destDir, 'lib', testPackage.moduleName)
       return Promise.all([
+        fs.copyAsync(path.join(__dirname, '../../assets'), path.join(moduleDir, 'assets')),
```
Comment (c-frater) [inline build/main/karma.js]: Added the assets to each lib to prevent `404` warnings for the icon font

File: build/main/karma.js [hunk 2]
```diff
@@ -48,7 +49,7 @@ module.exports = function (testPackages, destDir, phantomOnly) {
         ]
       })
     })).then(function (files) {
-      return runKarma(flatten(files), destDir, phantomOnly)
+      return runKarma(flatten([path.join(destDir, 'lib', 'test-helpers.js'), files]), destDir, chromeHeadlessOnly)
```
Comment (c-frater) [inline build/main/karma.js]: Include the test helpers once to reduce the number of `duplicate file` warnings

File: modules/autocomplete-feed/test/spec.coffee [hunk 3]
```diff
@@ -94,7 +94,7 @@ describe 'autocomplete-feed', ->
       af.items(items)
       callback = chai.spy()
       af.filter('a', callback)
-      callback.should.have.been.called.once()
+      callback.should.have.been.called.once
```
Comment (c-frater) [inline modules/autocomplete-feed/test/spec.coffee]: Updating Chai meant changing to use the updated API which does not treat these as functions

File: modules/autocomplete-picker/test/spec.coffee [hunk 4]
```diff
@@ -24,15 +29,22 @@ describe 'autocomplete-picker', ->
       callback(trivialItems)
     setTimeout(cb, trivialAsyncWait)
 
+  before ->
+    clock = sinon.useFakeTimers()
+    hx.loop = mock_hx_loop
```
Comment (c-frater) [inline modules/autocomplete-picker/test/spec.coffee]: The `open autocomplete` tests relied on the completion of the dropdown animation that was not correctly processed in the tests when using `useFakeTimers` so I copied the mock loop from the dropdown tests

File: modules/data-table/test/spec.coffee [hunk 5]
```diff
@@ -1224,7 +1241,7 @@ describe 'data-table', ->
         testTable {tableOptions}, done, (container, dt, options, data) ->
           container.select('.hx-sticky-table-wrapper').select('tbody').select('tr').selectAll('td').forEach (cell, index) ->
             if index is 0
-              cell.attr('style').should.equal('max-width: 10px; width: 10px; min-width: 10px; ')
+              cell.attr('style').should.equal('max-width: 10px; width: 10px; min-width: 10px;')
```
Comment (c-frater) [inline modules/data-table/test/spec.coffee]: The additional space here was one of the caveats for `phantomjs` that is no longer an issue

File: modules/date-localizer/main/index.coffee [hunk 6]
```diff
@@ -73,6 +73,10 @@ class DateTimeLocalizer extends PreferencesHandler
     if short then hx.zeroPad(month + 1)
     else ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sep','Oct','Nov','Dec'][month]
 
+  # localise the month in the format of mmm (Jan, Feb etc.) or 01, 02 etc.
+  fullMonth: (month) ->
+    ['January','February','March','April','May','June','July','August','September','October','November','December'][month]
```
Comment (c-frater) [inline modules/date-localizer/main/index.coffee]: This is one of the v2 features - The month displayed in the calendar has changed from `MMM / YYYY` to `MMMM YYYY` (e.g. from `Nov / 2018` to `November 2018`)

File: modules/date-localizer/test/spec.coffee [hunk 7]
```diff
@@ -17,6 +17,7 @@ describe 'dateTimeLocalizer', ->
   beforeEach ->
     hx.consoleWarning = chai.spy()
     localizer = hx.dateTimeLocalizer()
+    hx.preferences.timezone(hx._.preferences.defaultTimezoneLookup(0))
```
Comment (c-frater) [inline modules/date-localizer/test/spec.coffee]: The timezone changes in Chrome but not PhantomJS so we set it explicitly at the start of this test to prevent 'off by one' errors or problems when running in a different timezone

File: modules/date-picker/main/index.coffee [hunk 8]
```diff
@@ -92,39 +92,67 @@ isToday = (year, month, day) ->
...
+      if _.endDate < _.startDate
+        tDate = _.endDate
+
+        _.endDate = _.startDate
+        _.startDate = tDate
 
-  _.endDate.setHours(0, 0, 0, 0)
-  _.startDate.setHours(0, 0, 0, 0)
+    _.endDate.setHours(0, 0, 0, 0)
+    _.startDate.setHours(0, 0, 0, 0)
```
Comment (c-frater) [inline modules/date-picker/main/index.coffee]: We no longer modify the date values in this function for v2 and instead just show the date as invalid

File: modules/date-picker/main/index.coffee [hunk 9]
```diff
@@ -300,29 +331,32 @@ buildDatepicker = (datepicker) ->
...
     _.preventFeedback = true
     if datepicker.options.selectRange
       _.inputStart.classed('hx-date-error', false)
       _.inputEnd.classed('hx-date-error', false)
+    else if _.inputOnlyMode
+      datepicker.options.dateValidationChange?(true)
     else
       _.input.classed('hx-date-error', false)
-    setupInput datepicker
+    setupInput datepicker, initial
```
Comment (c-frater) [inline modules/date-picker/main/index.coffee]: When the datepicker is created, it should not set the `input` value so we use this `initial` flag to prevent it being set

File: modules/date-picker/main/index.coffee [hunk 10]
```diff
@@ -345,20 +379,23 @@ class DatePicker extends hx.EventEmitter
...
       defaultView: 'm' # 'm' for month, 'y' for year, or 'd' for decade
+      allowViewChange: true # Allow changing between month/year/decade views
       closeOnSelect: true
       selectRange: false
       validRange: undefined
       range: undefined
       showTodayButton: true
       allowInbuiltPicker: true # Option to allow preventing use of the inbuilt datepicker
       disabled: false
+      v2Features: false # Toggle all v2 functionality
```
Comment (c-frater) [inline modules/date-picker/main/index.coffee]: I'm going to split this up and make it an object with individual properties like

```js
v2Flags: {
  modifyDateOnError: true,
  displayLongMonthInCalendar: false,
  setInitialInputValue: true,
  updateVisibleMonthOnDateChange: false,
  allowInputElementContainer: false,
}
```

This allows us to make further v2 changes moving forwards if they are required without breaking the 1.x release for those using the flags to ease migration
